### ISSUE TITLE
synnom parser isn't compiling: unexpected end of macro invocation

### ISSUE THREAD

**Issue by chpio** [dtolnay/syn#99]
Im new to synnom/nom. The goal is to parse this example code:
```
    $ident,
    $ident<$path>,
    $ident: $ident,
    $ident {
        $ident: $ident,
    },
```

```rust
use syn::parse::{ident, path};
use syn::{Ident, Path};

named!(parse -> Vec<(Ident, Option<Path>, Option<Ident>, Vec<(Ident, Ident)>)>,
    terminated_list!(punct!(","), tuple!(
        ident,
        option!(delimited!(punct!("<"), path, punct!(">"))),
        option!(do_parse!(
            punct!(":") >>
            name: ident >>
            (name)
        )),
        opt_vec!(delimited!(
            punct!("{"),
            terminated_list!(punct!(","), do_parse!(
                name: ident >>
                punct!(":") >>
                ty_name: ident >>
                (name, ty_name)
            )),
            punct!("}")
        ))
    ))
);
```

But the parser isn't even compiling, rustc throws this error:

```
error: unexpected end of macro invocation
  --> .../src/parser.rs:24:31
   |
24 |                 (name, ty_name)
   |                               ^
```

Im using rustc version 1.17.0-nightly (60a0edc6c 2017-02-26) right now.

I have no idea if im doing something wrong, or if it's a bug in synnom. Thank you.

**Comment by dtolnay** [comment]
Your code is correct but terminated_list was only handling the most basic arguments:

> ```
> The implementation requires that the first parameter is a `punct!` macro,
> and the second is a named parser.
> ```

I extended it to support your use case in https://github.com/dtolnay/syn/commit/dff18805989ed5102f67d69a7afd0c043c4921a4 and published synom 0.11.2 with the fix. Thanks for reporting this!

**Comment by chpio** [comment]
> The implementation requires that the first parameter is a `punct!` macro,
and the second is a named parser.

so my code was actually broken.

> I extended it to support your use case in dff1880 and published synom 0.11.2 with the fix. Thanks for reporting this!

Thank you for adding this!
